File: examples/class-kits/gospider/ecommerce_profile.go
package classkit

import (
	"net/url"
	"regexp"
	"strings"
)

const defaultSiteFamily = "jd"

type ecommerceProfile struct {
	Family              string
	CatalogURL          string
	DetailURL           string
	ReviewURL           string
	Runner              string
	DetailLinkKeywords  []string
	NextLinkKeywords    []string
	ReviewLinkKeywords  []string
	PricePatterns       []string
	ItemIDPatterns      []string
	ShopPatterns        []string
	ReviewCountPatterns []string
	RatingPatterns      []string
}

func profileForFamily(siteFamily string) ecommerceProfile {
	switch strings.ToLower(siteFamily) {
	case "taobao":
		return ecommerceProfile{
			Family:              "taobao",
			CatalogURL:          "https://s.taobao.com/search?q=iphone",
			DetailURL:           "https://item.taobao.com/item.htm?id=100000000000",
			ReviewURL:           "https://rate.taobao.com/detailCommon.htm?id=100000000000",
			Runner:              "browser",
			DetailLinkKeywords:  []string{"item.taobao.com", "item.htm", "id=", "detail"},
			NextLinkKeywords:    []string{"page=", "next"},
			ReviewLinkKeywords:  []string{"review", "rate.taobao.com", "comment"},
			PricePatterns:       []string{`(?:price|promotionPrice|minPrice)["'=:\s]+([0-9]+(?:\.[0-9]{1,2})?)`, `(?:￥|¥)\s*([0-9]+(?:\.[0-9]{1,2})?)`},
			ItemIDPatterns:      []string{`(?:itemId|item_id|id)["'=:\s]+([A-Za-z0-9_-]+)`},
			ShopPatterns:        []string{`(?:shopName|sellerNick|nick)["'=:\s]+([^"'\\n<,}]+)`},
			ReviewCountPatterns: []string{`(?:reviewCount|commentCount|rateTotal)["'=:\s]+([0-9]+)`},
			RatingPatterns:      []string{`(?:score|rating)["'=:\s]+([0-9]+(?:\.[0-9])?)`},
		}
	case "tmall":
		return ecommerceProfile{
			Family:              "tmall",
			CatalogURL:          "https://list.tmall.com/search_product.htm?q=iphone",
			DetailURL:           "https://detail.tmall.com/item.htm?id=100000000000",
			ReviewURL:           "https://rate.tmall.com/list_detail_rate.htm?itemId=100000000000",
			Runner:              "browser",
			DetailLinkKeywords:  []string{"detail.tmall.com", "item.htm", "id=", "detail"},
			NextLinkKeywords:    []string{"page=", "next"},
			ReviewLinkKeywords:  []string{"review", "rate.tmall.com", "comment"},
			PricePatterns:       []string{`(?:price|promotionPrice|minPrice)["'=:\s]+([0-9]+(?:\.[0-9]{1,2})?)`, `(?:￥|¥)\s*([0-9]+(?:\.[0-9]{1,2})?)`},
			ItemIDPatterns:      []string{`(?:itemId|item_id|id)["'=:\s]+([A-Za-z0-9_-]+)`},
			ShopPatterns:        []string{`(?:shopName|sellerNick|shop)["'=:\s]+([^"'\\n<,}]+)`},
			ReviewCountPatterns: []string{`(?:reviewCount|commentCount|rateTotal)["'=:\s]+([0-9]+)`},
			RatingPatterns:      []string{`(?:score|rating)["'=:\s]+([0-9]+(?:\.[0-9])?)`},
		}
	case "pinduoduo":
		return ecommerceProfile{
			Family:              "pinduoduo",
			CatalogURL:          "https://mobile.yangkeduo.com/search_result.html?search_key=iphone",
			DetailURL:           "https://mobile.yangkeduo.com/goods.html?goods_id=100000000000",
			ReviewURL:           "https://mobile.yangkeduo.com/proxy/api/reviews/100000000000",
			Runner:              "browser",
			DetailLinkKeywords:  []string{"goods.html", "goods_id=", "product", "detail"},
			NextLinkKeywords:    []string{"page=", "next"},
			ReviewLinkKeywords:  []string{"review", "comment"},
			PricePatterns:       []string{`(?:minPrice|price|groupPrice)["'=:\s]+([0-9]+(?:\.[0-9]{1,2})?)`, `(?:￥|¥)\s*([0-9]+(?:\.[0-9]{1,2})?)`},
			ItemIDPatterns:      []string{`(?:goods_id|goodsId|skuId)["'=:\s]+([A-Za-z0-9_-]+)`},
			ShopPatterns:        []string{`(?:mall_name|storeName|shopName)["'=:\s]+([^"'\\n<,}]+)`},
			ReviewCountPatterns: []string{`(?:reviewCount|commentCount)["'=:\s]+([0-9]+)`},
			RatingPatterns:      []string{`(?:score|rating)["'=:\s]+([0-9]+(?:\.[0-9])?)`},
		}
	case "amazon":
		return ecommerceProfile{
			Family:              "amazon",
			CatalogURL:          "https://www.amazon.com/s?k=iphone",
			DetailURL:           "https://www.amazon.com/dp/B0EXAMPLE00",
			ReviewURL:           "https://www.amazon.com/product-reviews/B0EXAMPLE00",
			Runner:              "browser",
			DetailLinkKeywords:  []string{"/dp/", "/gp/product/", "/product/", "asin"},
			NextLinkKeywords:    []string{"page=", "next"},
			ReviewLinkKeywords:  []string{"review", "product-reviews"},
			PricePatterns:       []string{`(?:priceToPay|displayPrice|priceAmount)["'=:\s]+([0-9]+(?:\.[0-9]{1,2})?)`, `\$\s*([0-9]+(?:\.[0-9]{1,2})?)`},
			ItemIDPatterns:      []string{`(?:asin|parentAsin|sku)["'=:\s]+([A-Za-z0-9_-]+)`},
			ShopPatterns:        []string{`(?:seller|merchantName|bylineInfo)["'=:\s]+([^"'\\n<,}]+)`},
			ReviewCountPatterns: []string{`(?:reviewCount|totalReviewCount)["'=:\s]+([0-9]+)`},
			RatingPatterns:      []string{`(?:averageRating|rating)["'=:\s]+([0-9]+(?:\.[0-9])?)`},
		}
	default:
		return ecommerceProfile{
			Family:              "jd",
			CatalogURL:          "https://search.jd.com/Search?keyword=iphone",
			DetailURL:           "https://item.jd.com/100000000000.html",
			ReviewURL:           "https://club.jd.com/comment/productPageComments.action?productId=100000000000",
			Runner:              "browser",
			DetailLinkKeywords:  []string{"item.jd.com", "sku=", "wareId=", "item.htm", "detail"},
			NextLinkKeywords:    []string{"page=", "pn-next", "next"},
			ReviewLinkKeywords:  []string{"comment", "review", "club.jd.com"},
			PricePatterns:       []string{`"p"\s*:\s*"([0-9]+(?:\.[0-9]{1,2})?)"`, `(?:price|jdPrice|promotionPrice)["'=:\s]+([0-9]+(?:\.[0-9]{1,2})?)`, `(?:￥|¥)\s*([0-9]+(?:\.[0-9]{1,2})?)`},
			ItemIDPatterns:      []string{`(?:skuId|sku|wareId|productId)["'=:\s]+([A-Za-z0-9_-]+)`},
			ShopPatterns:        []string{`(?:shopName|venderName|storeName)["'=:\s]+([^"'\\n<,}]+)`},
			ReviewCountPatterns: []string{`(?:commentCount|comment_num|reviewCount)["'=:\s]+([0-9]+)`},
			RatingPatterns:      []string{`(?:score|rating)["'=:\s]+([0-9]+(?:\.[0-9])?)`},
		}
	}
}

func firstRegexMatch(text string, patterns []string) string {
	for _, pattern := range patterns {
		re := regexp.MustCompile(`(?i)` + pattern)
		match := re.FindStringSubmatch(text)
		if len(match) > 1 {
			return strings.TrimSpace(match[1])
		}
	}
	return ""
}

func collectRegexMatches(text string, patterns []string, limit int) []string {
	values := make([]string, 0, limit)
	seen := map[string]bool{}
	for _, pattern := range patterns {
		re := regexp.MustCompile(`(?i)` + pattern)
		matches := re.FindAllStringSubmatch(text, -1)
		for _, match := range matches {
			if len(match) < 2 {
				continue
			}
			value := strings.TrimSpace(match[1])
			if value == "" || seen[value] {
				continue
			}
			seen[value] = true
			values = append(values, value)
			if len(values) >= limit {
				return values
			}
		}
	}
	return values
}

func normalizeLinks(baseURL string, links []string) []string {
	values := make([]string, 0, len(links))
	seen := map[string]bool{}
	base, _ := url.Parse(baseURL)
	for _, link := range links {
		link = strings.TrimSpace(link)
		if link == "" {
			continue
		}
		absolute := link
		if base != nil {
			if ref, err := url.Parse(link); err == nil {
				absolute = base.ResolveReference(ref).String()
			}
		}
		if !strings.HasPrefix(absolute, "http://") && !strings.HasPrefix(absolute, "https://") {
			continue
		}
		if !seen[absolute] {
			seen[absolute] = true
			values = append(values, absolute)
		}
	}
	return values
}

func collectProductLinks(baseURL string, links []string, profile ecommerceProfile, limit int) []string {
	values := make([]string, 0, limit)
	for _, link := range normalizeLinks(baseURL, links) {
		lowered := strings.ToLower(link)
		for _, keyword := range profile.DetailLinkKeywords {
			if strings.Contains(lowered, strings.ToLower(keyword)) {
				values = append(values, link)
				break
			}
		}
		if len(values) >= limit {
			break
		}
	}
	return values
}

func collectImageLinks(baseURL string, links []string, limit int) []string {
	values := make([]string, 0, limit)
	for _, link := range normalizeLinks(baseURL, links) {
		lowered := strings.ToLower(link)
		if strings.Contains(lowered, "image") ||
			strings.HasSuffix(lowered, ".jpg") ||
			strings.HasSuffix(lowered, ".jpeg") ||
			strings.HasSuffix(lowered, ".png") ||
			strings.HasSuffix(lowered, ".webp") ||
			strings.HasSuffix(lowered, ".gif") {
			values = append(values, link)
		}
		if len(values) >= limit {
			break
		}
	}
	return values
}

func firstLinkWithKeywords(baseURL string, links []string, keywords []string) string {
	for _, link := range normalizeLinks(baseURL, links) {
		lowered := strings.ToLower(link)
		for _, keyword := range keywords {
			if strings.Contains(lowered, strings.ToLower(keyword)) {
				return link
			}
		}
	}
	return ""
}

func bestTitle(responseTitle string, h1 string) string {
	if strings.TrimSpace(responseTitle) != "" {
		return strings.TrimSpace(responseTitle)
	}
	return strings.TrimSpace(h1)
}

func excerpt(text string, limit int) string {
	normalized := strings.Join(strings.Fields(text), " ")
	if len(normalized) <= limit {
		return normalized
	}
	return normalized[:limit]
}

func siteFamilyFromResponse(responseURL string, requestMeta map[string]any) string {
	if requestMeta != nil {
		if raw, ok := requestMeta["site_family"].(string); ok && raw != "" {
			return raw
		}
	}
	lowered := strings.ToLower(responseURL)
	switch {
	case strings.Contains(lowered, "taobao.com"):
		return "taobao"
	case strings.Contains(lowered, "tmall.com"):
		return "tmall"
	case strings.Contains(lowered, "yangkeduo.com") || strings.Contains(lowered, "pinduoduo.com"):
		return "pinduoduo"
	case strings.Contains(lowered, "amazon.com"):
		return "amazon"
	default:
		return defaultSiteFamily
	}
}

File: examples/class-kits/rustspider/ecommerce_profile.rs
use regex::Regex;
use rustspider::scrapy::Response;
use serde_json::Value;
use url::Url;

pub const DEFAULT_SITE_FAMILY: &str = "jd";

#[derive(Clone)]
pub struct EcommerceProfile {
    pub family: &'static str,
    pub catalog_url: &'static str,
    pub detail_url: &'static str,
    pub review_url: &'static str,
    pub runner: &'static str,
    pub detail_link_keywords: &'static [&'static str],
    pub next_link_keywords: &'static [&'static str],
    pub review_link_keywords: &'static [&'static str],
    pub price_patterns: &'static [&'static str],
    pub item_id_patterns: &'static [&'static str],
    pub shop_patterns: &'static [&'static str],
    pub review_count_patterns: &'static [&'static str],
    pub rating_patterns: &'static [&'static str],
}

pub fn profile_for_family(site_family: &str) -> EcommerceProfile {
    match site_family.to_ascii_lowercase().as_str() {
        "taobao" => EcommerceProfile {
            family: "taobao",
            catalog_url: "https://s.taobao.com/search?q=iphone",
            detail_url: "https://item.taobao.com/item.htm?id=100000000000",
            review_url: "https://rate.taobao.com/detailCommon.htm?id=100000000000",
            runner: "browser",
            detail_link_keywords: &["item.taobao.com", "item.htm", "id=", "detail"],
            next_link_keywords: &["page=", "next"],
            review_link_keywords: &["review", "rate.taobao.com", "comment"],
            price_patterns: &[
                r#"(?:price|promotionPrice|minPrice)["'=:\s]+([0-9]+(?:\.[0-9]{1,2})?)"#,
                r#"(?:￥|¥)\s*([0-9]+(?:\.[0-9]{1,2})?)"#,
            ],
            item_id_patterns: &[r#"(?:itemId|item_id|id)["'=:\s]+([A-Za-z0-9_-]+)"#],
            shop_patterns: &[r#"(?:shopName|sellerNick|nick)["'=:\s]+([^"'\n<,}]+)"#],
            review_count_patterns: &[r#"(?:reviewCount|commentCount|rateTotal)["'=:\s]+([0-9]+)"#],
            rating_patterns: &[r#"(?:score|rating)["'=:\s]+([0-9]+(?:\.[0-9])?)"#],
        },
        "tmall" => EcommerceProfile {
            family: "tmall",
            catalog_url: "https://list.tmall.com/search_product.htm?q=iphone",
            detail_url: "https://detail.tmall.com/item.htm?id=100000000000",
            review_url: "https://rate.tmall.com/list_detail_rate.htm?itemId=100000000000",
            runner: "browser",
            detail_link_keywords: &["detail.tmall.com", "item.htm", "id=", "detail"],
            next_link_keywords: &["page=", "next"],
            review_link_keywords: &["review", "rate.tmall.com", "comment"],
            price_patterns: &[
                r#"(?:price|promotionPrice|minPrice)["'=:\s]+([0-9]+(?:\.[0-9]{1,2})?)"#,
                r#"(?:￥|¥)\s*([0-9]+(?:\.[0-9]{1,2})?)"#,
            ],
            item_id_patterns: &[r#"(?:itemId|item_id|id)["'=:\s]+([A-Za-z0-9_-]+)"#],
            shop_patterns: &[r#"(?:shopName|sellerNick|shop)["'=:\s]+([^"'\n<,}]+)"#],
            review_count_patterns: &[r#"(?:reviewCount|commentCount|rateTotal)["'=:\s]+([0-9]+)"#],
            rating_patterns: &[r#"(?:score|rating)["'=:\s]+([0-9]+(?:\.[0-9])?)"#],
        },
        "pinduoduo" => EcommerceProfile {
            family: "pinduoduo",
            catalog_url: "https://mobile.yangkeduo.com/search_result.html?search_key=iphone",
            detail_url: "https://mobile.yangkeduo.com/goods.html?goods_id=100000000000",
            review_url: "https://mobile.yangkeduo.com/proxy/api/reviews/100000000000",
            runner: "browser",
            detail_link_keywords: &["goods.html", "goods_id=", "product", "detail"],
            next_link_keywords: &["page=", "next"],
            review_link_keywords: &["review", "comment"],
            price_patterns: &[
                r#"(?:minPrice|price|groupPrice)["'=:\s]+([0-9]+(?:\.[0-9]{1,2})?)"#,
                r#"(?:￥|¥)\s*([0-9]+(?:\.[0-9]{1,2})?)"#,
            ],
            item_id_patterns: &[r#"(?:goods_id|goodsId|skuId)["'=:\s]+([A-Za-z0-9_-]+)"#],
            shop_patterns: &[r#"(?:mall_name|storeName|shopName)["'=:\s]+([^"'\n<,}]+)"#],
            review_count_patterns: &[r#"(?:reviewCount|commentCount)["'=:\s]+([0-9]+)"#],
            rating_patterns: &[r#"(?:score|rating)["'=:\s]+([0-9]+(?:\.[0-9])?)"#],
        },
        "amazon" => EcommerceProfile {
            family: "amazon",
            catalog_url: "https://www.amazon.com/s?k=iphone",
            detail_url: "https://www.amazon.com/dp/B0EXAMPLE00",
            review_url: "https://www.amazon.com/product-reviews/B0EXAMPLE00",
            runner: "browser",
            detail_link_keywords: &["/dp/", "/gp/product/", "/product/", "asin"],
            next_link_keywords: &["page=", "next"],
            review_link_keywords: &["review", "product-reviews"],
            price_patterns: &[
                r#"(?:priceToPay|displayPrice|priceAmount)["'=:\s]+([0-9]+(?:\.[0-9]{1,2})?)"#,
                r#"\$\s*([0-9]+(?:\.[0-9]{1,2})?)"#,
            ],
            item_id_patterns: &[r#"(?:asin|parentAsin|sku)["'=:\s]+([A-Za-z0-9_-]+)"#],
            shop_patterns: &[r#"(?:seller|merchantName|bylineInfo)["'=:\s]+([^"'\n<,}]+)"#],
            review_count_patterns: &[r#"(?:reviewCount|totalReviewCount)["'=:\s]+([0-9]+)"#],
            rating_patterns: &[r#"(?:averageRating|rating)["'=:\s]+([0-9]+(?:\.[0-9])?)"#],
        },
        _ => EcommerceProfile {
            family: "jd",
            catalog_url: "https://search.jd.com/Search?keyword=iphone",
            detail_url: "https://item.jd.com/100000000000.html",
            review_url:
                "https://club.jd.com/comment/productPageComments.action?productId=100000000000",
            runner: "browser",
            detail_link_keywords: &["item.jd.com", "sku=", "wareId=", "item.htm", "detail"],
            next_link_keywords: &["page=", "pn-next", "next"],
            review_link_keywords: &["comment", "review", "club.jd.com"],
            price_patterns: &[
                r#""p"\s*:\s*"([0-9]+(?:\.[0-9]{1,2})?)""#,
                r#"(?:price|jdPrice|promotionPrice)["'=:\s]+([0-9]+(?:\.[0-9]{1,2})?)"#,
                r#"(?:￥|¥)\s*([0-9]+(?:\.[0-9]{1,2})?)"#,
            ],
            item_id_patterns: &[r#"(?:skuId|sku|wareId|productId)["'=:\s]+([A-Za-z0-9_-]+)"#],
            shop_patterns: &[r#"(?:shopName|venderName|storeName)["'=:\s]+([^"'\n<,}]+)"#],
            review_count_patterns: &[
                r#"(?:commentCount|comment_num|reviewCount)["'=:\s]+([0-9]+)"#,
            ],
            rating_patterns: &[r#"(?:score|rating)["'=:\s]+([0-9]+(?:\.[0-9])?)"#],
        },
    }
}

pub fn first_match(text: &str, patterns: &[&str]) -> String {
    for pattern in patterns {
        if let Ok(regex) = Regex::new(&format!("(?i){pattern}")) {
            if let Some(capture) = regex.captures(text) {
                if let Some(value) = capture.get(1) {
                    return value.as_str().trim().to_string();
                }
            }
        }
    }
    String::new()
}

pub fn collect_matches(text: &str, patterns: &[&str], limit: usize) -> Vec<String> {
    let mut values = Vec::new();
    for pattern in patterns {
        if let Ok(regex) = Regex::new(&format!("(?i){pattern}")) {
            for capture in regex.captures_iter(text) {
                if let Some(value) = capture.get(1) {
                    let candidate = value.as_str().trim().to_string();
                    if !candidate.is_empty() && !values.contains(&candidate) {
                        values.push(candidate);
                    }
                    if values.len() >= limit {
                        return values;
                    }
                }
            }
        }
    }
    values
}

pub fn normalize_links(base_url: &str, links: Vec<String>) -> Vec<String> {
    let mut values = Vec::new();
    let base = Url::parse(base_url).ok();
    for link in links {
        let candidate = link.trim();
        if candidate.is_empty() {
            continue;
        }
        let absolute = match (base.as_ref(), Url::parse(candidate).ok()) {
            (_, Some(url)) => url.to_string(),
            (Some(base_url), None) => base_url
                .join(candidate)
                .ok()
                .map(|url| url.to_string())
                .unwrap_or_default(),
            (None, None) => String::new(),
        };
        if absolute.starts_with("http://") || absolute.starts_with("https://") {
            if !values.contains(&absolute) {
                values.push(absolute);
            }
        }
    }
    values
}

pub fn collect_product_links(
    base_url: &str,
    links: Vec<String>,
    profile: &EcommerceProfile,
    limit: usize,
) -> Vec<String> {
    let mut values = Vec::new();
    for link in normalize_links(base_url, links) {
        let lowered = link.to_ascii_lowercase();
        if profile
            .detail_link_keywords
            .iter()
            .any(|keyword| lowered.contains(&keyword.to_ascii_lowercase()))
        {
            values.push(link);
        }
        if values.len() >= limit {
            break;
        }
    }
    values
}

pub fn collect_image_links(base_url: &str, links: Vec<String>, limit: usize) -> Vec<String> {
    let mut values = Vec::new();
    for link in normalize_links(base_url, links) {
        let lowered = link.to_ascii_lowercase();
        if lowered.contains("image")
            || lowered.ends_with(".jpg")
            || lowered.ends_with(".jpeg")
            || lowered.ends_with(".png")
            || lowered.ends_with(".webp")
            || lowered.ends_with(".gif")
        {
            values.push(link);
        }
        if values.len() >= limit {
            break;
        }
    }
    values
}

pub fn first_link_with_keywords(base_url: &str, links: Vec<String>, keywords: &[&str]) -> String {
    for link in normalize_links(base_url, links) {
        let lowered = link.to_ascii_lowercase();
        if keywords
            .iter()
            .any(|keyword| lowered.contains(&keyword.to_ascii_lowercase()))
        {
            return link;
        }
    }
    String::new()
}

pub fn best_title(response: &Response) -> String {
    response
        .css("title")
        .get()
        .or_else(|| response.css("h1").get())
        .unwrap_or_default()
}

pub fn site_family_from_response(response: &Response) -> String {
    if let Some(request) = &response.request {
        if let Some(Value::String(raw)) = request.meta.get("site_family") {
            if !raw.is_empty() {
                return raw.to_string();
            }
        }
    }
    let lowered = response.url.to_ascii_lowercase();
    if lowered.contains("taobao.com") {
        return "taobao".to_string();
    }
    if lowered.contains("tmall.com") {
        return "tmall".to_string();
    }
    if lowered.contains("yangkeduo.com") || lowered.contains("pinduoduo.com") {
        return "pinduoduo".to_string();
    }
    if lowered.contains("amazon.com") {
        return "amazon".to_string();
    }
    DEFAULT_SITE_FAMILY.to_string()
}

pub fn text_excerpt(text: &str, limit: usize) -> String {
    let normalized = text.split_whitespace().collect::<Vec<_>>().join(" ");
    normalized.chars().take(limit).collect()
}
